### ISSUE TITLE
nodatime types doesn't work with minimal apis when used as a query parameter

### ISSUE THREAD

**Issue by juantxorena** [nodatime/nodatime#1833]
As the title says, using nodatime types as query parameters on a [minimal api](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/overview) doesn't compile, it throws the following error:

`ASP0020	Parameter 'datetime' of type Instant should define a bool TryParse(string, IFormatProvider, out Instant) method, or implement IParsable<Instant>`

Here's an example of an endpoint with this problem:

```
var builder = WebApplication.CreateBuilder(args);
builder.Services.ConfigureHttpJsonOptions(options =>
{
        options.SerializerOptions.ConfigureForNodaTime(DateTimeZoneProviders.Tzdb);
});
var app = builder.Build();
app.MapGet("realtimedata/now/{vehicleId}", (Guid vehicleId, [FromQuery] Instant datetime) => $"vehicle id: {vehicleId}, datetime: {datetime}");

app.Run();

```

If I put the Instant in the body or directly in the path of the query, it seems to work (at least it doesn't throw an error):
```
var builder = WebApplication.CreateBuilder(args);
builder.Services.ConfigureHttpJsonOptions(options =>
{
        options.SerializerOptions.ConfigureForNodaTime(DateTimeZoneProviders.Tzdb);
});
var app = builder.Build();
app.MapGet("realtimedata/now/{vehicleId}/{datetime}", (Guid vehicleId, Instant datetime) => $"vehicle id: {vehicleId}, datetime: {datetime}");

app.Run();

```

**Comment by jskeet** [comment]
I don't know the details of how minimal APIs expect to use `[FromQuery]` parameters, and don't have the time to research that at the moment. If you can find the relevant documentation for how that can be customized, I'm happy to look at at it a bit more, but fundamentally Noda Time provides JSON serialization libaries for both Json.NET and System.Text.Json - I would really have expected that to be enough. We could *potentially* support `IParsable<T>` and/or provide a `TryParse` method, but it feels like we'll endlessly be chasing our tales if we're not careful.

**Comment by jskeet** [comment]
How frustrating. I'll have a think about possibly supporting `IParsable<T>`. It seems odd that there isn't even a way of saying what format is expecting (e.g. "dd/MM/yyyy" vs "yyyy-MM-dd") other than with the `IFormatProvider`. I suspect I could have some horribly hacky way of having a placeholder between Noda Time's `IPattern<T>` and `IFormatProvider`, such that if you call `formatProvider.Get(typeof(IPattern<Instant>))` for example, it returns the original `IPattern<Instant>` that it was created from. (I could even make `IPattern<T>` implement `IFormatProvider` itself...) But I'm not at all *convinced* that's a good idea.

Any other ideas would be welcome :)

**Comment by juantxorena** [comment]
The docs also talk about implementing BindAsync. I have no idea about any of this, so I cannot give any opinion:

https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/parameter-binding?view=aspnetcore-8.0#custom-binding

**Comment by nandor23** [comment]
The issue persists, in the meantime you could write your own parser. I am using FastEndpoints, but it works just like minimal APIs. You can use the `InstantPattern.ExtendedIso.Parse(value);` to parse from `string` to `Instant`.

I will provide my code for reference, maybe it would be useful for someone:
```c#
    public static ParseResult InstantParser(object? input)
    {
        var value = input?.ToString();

        if (value is null)
        {
            return new ParseResult(false, new Instant());
        }

        try
        {
            var result = InstantPattern.ExtendedIso.Parse(value);
            return new ParseResult(result.Success, result.Value);
        }
        catch
        {
            return new ParseResult(false, new Instant());
        }
    }
```

The return type `ParseResult` is FastEndpoint specific, not the same as the default .NET one.

**Comment by dukesteen** [comment]
@jskeet Did you have any more thoughts about this issue? It's still a very annoying issue with minimal API's

**Comment by jskeet** [comment]
@dukesteen: No, I haven't looked at it at all. I'm still open to hearing about concrete ideas for the best way to implement this.

**Comment by jtnicholl** [comment]
There are some related interfaces that Microsoft has started implementing for all the core types, which allow for a lot more performant code.

[`ISpanParsable<TSelf>`](https://learn.microsoft.com/en-us/dotnet/api/system.ispanparsable-1) allows parsing from char spans, which are slices of existing strings or arrays ([or anything else, if you're brave enough](https://learn.microsoft.com/en-us/dotnet/api/system.span-1.-ctor?view=net-10.0#system-span-1-ctor(system-void*-system-int32))). Spans never require any heap allocations, so they are good for performance. `ISpanParsable<TSelf>` extends `IParsable<TSelf>`, I assume because creating a span over a string is a trivial operation. There is a corresponding [`ISpanFormattable`](https://learn.microsoft.com/en-us/dotnet/api/system.ispanformattable) which serializes to an existing buffer ([which can be on the stack](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/stackalloc)).

In .NET 8 they also added [`IUtf8SpanParsable<TSelf>`](https://learn.microsoft.com/en-us/dotnet/api/system.iutf8spanparsable-1) and [`IUtf8SpanFormattable`](https://learn.microsoft.com/en-us/dotnet/api/system.iutf8spanformattable), which can improve performance further by operating on byte spans and skipping conversion between UTF-8 and UTF-16 (since .NET strings are stored in UTF-16).

I would expect that in the near future, query parameter binding and a lot of other features will look for the UTF-8 interfaces first, then fall back to the char span and string interfaces in that order. (I think I remember reading someone at Microsoft say that was the plan, but I can't find it now.)

**Comment by jskeet** [comment]
@jtnicholl That's good to know, thanks. It's probably challenging to add parsing from `ReadOnlySpan<char>` - I'm sure I've tried it in the past and got stuck, but that was quite a while ago.

I have no problem adding code that only lights up on .NET 8. I have some free time coming up soon, so I may see what I can do.

**Comment by chhaugen** [comment]
This is my workaround for Fast Enpoints based on the System.Text.Json serializer.
[Add Value Parser For NodaTime based on NodaTime.Serialization.SystemTextJson](https://gist.github.com/chhaugen/65211f6d0d0bd58f03935da0e74b8ead)

Thanks to the Nodatime folks for the library, and thanks to @nandor23 for the inspiration for the workaround.
